// @sap/cds/bin/build/provider/buildTaskProviderInternal.js
const fs = require('fs')
const path = require('path')
const cds = require('../cds')
const { hasJavaNature, getProperty, isStreamlinedMtx, getDefaultModelOptions, BuildError } = require('../util')
const BuildTaskProvider = require('../buildTaskProvider')

const { FILE_EXT_CDS, BUILD_TASK_HANA, BUILD_TASK_FIORI, BUILD_TASK_JAVA, BUILD_TASK_JAVA_CF, BUILD_TASK_NODEJS, BUILD_TASK_NODE_CF, BUILD_TASK_MTX,
    BUILD_TASK_PREFIX, BUILD_TASKS, BUILD_TASK_MTX_SIDECAR, MTX_SIDECAR_FOLDER, BUILD_TASK_MTX_EXTENSION, CDS_MODEL_EXCLUDE_LIST } = require("../constants")

class BuildTaskProviderInternal extends BuildTaskProvider {
    constructor(logger) {
        super()
        this._logger = logger
    }
    get logger() {
        return this._logger
    }

    canHandleTask(task) {
        return BUILD_TASKS.includes(task.for)
            || task.use && task.use.startsWith(BUILD_TASK_PREFIX)
        // support different build task implementations of the same build task ID defined by 'for'
        //return (BUILD_TASKS.includes(task.for) && !task.use) || (task.use && task.use.startsWith(BUILD_TASK_PREFIX))
    }

    loadHandler(task) {
        return require(`./${BuildTaskProviderInternal._getForValueFromTask(task)}`)
    }

    async lookupTasks(tasks, buildOptions) {
        return this._createTasks(tasks, buildOptions, tasks.length > 0)
    }

    async applyTaskDefaults(task, buildOptions) {
        const taskFor = BuildTaskProviderInternal._getForValueFromTask(task)
        const { root: projectPath } = buildOptions
        task.for = task.for || taskFor
        task.use = task.use || `${BUILD_TASK_PREFIX}/${taskFor}`

        switch (taskFor) {
            case BUILD_TASK_HANA:
                task.src = task.src || BuildTaskProviderInternal._normalizePath(cds.env.folders.db)
                break
            case BUILD_TASK_JAVA:
            case BUILD_TASK_JAVA_CF:
                task.src = task.src || BuildTaskProviderInternal._normalizePath(cds.env.folders.srv)
                if (!task.options?.model) {
                    BuildTaskProviderInternal._setDefaultModelOptionsForJava(task, projectPath)
                }
                break
            case BUILD_TASK_NODEJS:
            case BUILD_TASK_NODE_CF:
                task.src = task.src || BuildTaskProviderInternal._normalizePath(cds.env.folders.srv)
                break
            case BUILD_TASK_FIORI:
                task.src = task.src || BuildTaskProviderInternal._normalizePath(cds.env.folders.app)
                break
            case BUILD_TASK_MTX_SIDECAR:
                task.src = task.src || MTX_SIDECAR_FOLDER
                if (!task.options?.model && BuildTaskProviderInternal._hasJavaNature(projectPath, cds.env.folders.srv)) {
                    BuildTaskProviderInternal._setDefaultModelOptionsForJava(task, projectPath)
                }
                break
            case BUILD_TASK_MTX_EXTENSION:
            case BUILD_TASK_MTX:
                if (isStreamlinedMtx()) {
                    task.src = task.src || BuildTaskProviderInternal._normalizePath(cds.env.folders.srv)
                } else {
                    task.src = task.src || "."
                }
                break
            default:
                throw new Error(`Unknown build task '${task.use || task.for}'`)
        }
    }

    async _createTasks(tasks, buildOptions, addRequiredTasks) {
        const { root: projectPath } = buildOptions
        let db = typeof cds.env.folders.db === "string" ? [BuildTaskProviderInternal._normalizePath(cds.env.folders.db)] : cds.env.folders.db
        let srv = typeof cds.env.folders.srv === "string" ? [BuildTaskProviderInternal._normalizePath(cds.env.folders.srv)] : cds.env.folders.srv

        const dbOptions = {
            model: []
        }
        const srvOptions = {
            model: []
        }
        if (Array.isArray(db) && db.length > 0) {
            db = BuildTaskProviderInternal._getModuleFolder(projectPath, db) || null
            if (!db && !addRequiredTasks) { // log once
                this.logger.debug("No database module found")
            }
        }
        if (Array.isArray(srv) && srv.length > 0) {
            srv = BuildTaskProviderInternal._getModuleFolder(projectPath, srv) || null
            if (!srv && !addRequiredTasks) { // log once
                this.logger.debug("No service module found")
            }
        }
        // create required build tasks
        if (addRequiredTasks) {
            const mtxTask = tasks.find(task => task.for === BUILD_TASK_MTX || task.for === BUILD_TASK_MTX_SIDECAR)
            if (mtxTask) {
                //restore used tasks as they might have been filtered by 'cds build --for mtx'
                if (mtxTask._uses) {
                    mtxTask._uses.map(use => use.task).forEach(task => {
                        if (!tasks.includes(task)) {
                            tasks.push(task)
                        }
                    })
                }
                let dbTask = tasks.find(task => task.for === BUILD_TASK_HANA)
                if (!dbTask) {
                    //  db task might be missing if mtx task is enforced by 'cds build --for mtx'
                    dbTask = this._createDbTask(projectPath, db, dbOptions)
                    if (dbTask) {
                        tasks.push(dbTask)
                    }
                }
                if (dbTask && (!mtxTask._uses || !mtxTask._uses.map(use => use.task).includes(dbTask))) {
                    const use = { task: dbTask }
                    mtxTask._uses ? mtxTask._uses.push(use) : mtxTask._uses = [use]
                }
            }
        } else {
            !db && this.logger._debug && this.logger.debug(`project doesn't have a database module '${cds.env.folders.db}'`)
            !srv && this.logger._debug && this.logger.debug(`project doesn't have a service module '${cds.env.folders.srv}'`)

            // create hana build task
            const dbTask = this._createDbTask(projectPath, db, dbOptions, buildOptions)
            if (dbTask) {
                tasks.push(dbTask)
            }
            // create java or node build task
            const srvTask = this._createSrvTask(projectPath, srv, srvOptions)
            if (srvTask) {
                tasks.push(srvTask)
            }
            // create mtx build task
            const mtxTask = this._createMtxTask(projectPath, srv, srv, tasks)
            if (mtxTask) {
                tasks.push(mtxTask)
            }
        }
    }

    _createDbTask(projectPath, src, taskOptions, buildOptions) {
        this.logger.debug("determining database kind.")
        if (!src || BuildTaskProviderInternal._isMtxExtension()) {
            return null
        }
        let task = null
        if (this._useHana(projectPath, buildOptions)) {
            this.logger.debug("found HANA database.")
            // legacy build supports dest property
            const compileDest = cds.env.data?.dest
            if (compileDest) {
                //../db/src/gen
                // compileDest is relative to src folder in modular build - resolve correctly
                taskOptions.compileDest = path.relative(path.resolve(projectPath, src), path.resolve(projectPath, compileDest))
            }
            task = {
                src: src,
                for: BUILD_TASK_HANA,
                options: taskOptions
            }
        } else {
            this.logger.debug("found sqlite database - skipping HANA build task")
        }
        return task
    }

    _useHana(projectPath, buildOptions) {
        if (getProperty(buildOptions, "for.hana.skipManifestGeneration") // deprecated fallback for webide fullstack and mtx
            || getProperty(buildOptions, "for.hana.contentManifest") === false // fallback for webide fullstack and mtx
            || cds.env.requires.db?.kind === "hana"
            || cds.env.requires.db?.dialect === "hana") {

            return true
        }
        // false if other db has been defined
        if (cds.env.requires.db?.kind) {
            return false
        }
        // check whether cds config represents a legacy build system config for which requires.db was not configured
        // Note: compat layer sets requires.db: {}
        const userEnv = cds.env.for("cds", projectPath, false)
        return userEnv && (userEnv.data?.model || userEnv.service?.model)
    }

    _createMtxTask(projectPath, src, dest, tasks) {
        // MTX build task creation is NOT supported for Java projects
        if (tasks.some(task => task.for === BUILD_TASK_JAVA || task.for === BUILD_TASK_JAVA_CF)) {
            return null
        }
        this.logger.debug("determining mtx version of nodejs project")

        // preserve order of creation
        if (BuildTaskProviderInternal._isMtxExtension()) {
            this.logger.debug("Streamlined MTX extension app")
            return {
                for: BUILD_TASK_MTX_EXTENSION
            }
        }

        if (isStreamlinedMtx()) {
            if (cds.env.requires["cds.xt.ModelProviderService"]?.kind === "rest") { // "cds.xt.ModelProviderService": "from-sidecar"
                this.logger.debug("Nodejs Streamlined MTX app with sidecar")

                const sidecarPath = path.join(projectPath, MTX_SIDECAR_FOLDER)
                if (!fs.existsSync(sidecarPath)) {
                    throw new BuildError(`MTX sidecar directory '${sidecarPath}' not existing. Custom build task configuration necessary if the folder is named differently.`)
                }
                const sidecarEnv = cds.env.for("cds", sidecarPath)
                if (sidecarEnv.requires["cds.xt.ModelProviderService"]?.kind === "in-sidecar") {
                    return {
                        for: BUILD_TASK_MTX_SIDECAR
                    }
                }
                throw new BuildError("Invalid MTX sidecar configuration - \"cds.xt.ModelProviderService\": \"in-sidecar\" missing.")
            }

            if (cds.env.requires["cds.xt.ModelProviderService"]?.kind === "in-sidecar") {
                // cds build is executed in sidecar folder
                throw new BuildError("Invalid working directory. Make sure to execute 'cds build' in CAP project root directory.")
            }

            this.logger.debug("Nodejs Streamlined MTX app without sidecar")
            return {
                for: BUILD_TASK_MTX,
                src
            }
        }

        if (cds.env.requires.multitenancy) {
            this.logger.debug("Nodejs Classic MTX app without sidecar")
            return {
                src: ".",
                for: BUILD_TASK_MTX,
                dest
            }
        }
    }

    _createSrvTask(projectPath, src, taskOptions) {
        this.logger.debug("determining implementation technology")
        if (!src || BuildTaskProviderInternal._isMtxExtension()) {
            return null
        }
        let task = null
        if (BuildTaskProviderInternal._hasJavaNature(projectPath, src)) {
            task = this._createJavaTask(projectPath, src, taskOptions)
        } else {
            task = this._createNodeTask(src, taskOptions)
        }
        return task
    }

    _createJavaTask(projectPath, src, taskOptions) {
        this.logger.debug("found implementation technology java")
        // legacy build supports dest property
        const compileDest = cds.env.service?.dest
        if (compileDest) {
            // compileDest is relative to src folder in modular build - resolve correctly
            taskOptions.compileDest = path.relative(path.resolve(projectPath, src), path.resolve(projectPath, compileDest))
        }
        return {
            src: src,
            for: BUILD_TASK_JAVA,
            options: taskOptions
        }
    }

    _createNodeTask(src, taskOptions) {
        this.logger.debug("found implementation technology node")
        return {
            src: src,
            for: BUILD_TASK_NODEJS,
            options: taskOptions
        }
    }

    static _isMtxExtension() {
        return !!cds.env.extends
    }

    /**
     * REVISIT: filter model options in order to avoid that the bootstrap service gets added to the application model
     * https://github.tools.sap/cap/issues/issues/12770#issuecomment-1805719
     */
    static _setDefaultModelOptionsForJava(task, projectPath) {
        const defaultModelPaths = new Set(getDefaultModelOptions(projectPath))
        defaultModelPaths.add(task.src)
        CDS_MODEL_EXCLUDE_LIST.forEach(m => defaultModelPaths.delete(m))
        task.options = task.options || {}
        task.options.model = [...defaultModelPaths]
    }

    /**
     * Returns whether this project is a java project or not.
     * @param {string} projectPath - the absolute project path
     * @param {string} src - the folder name of the service module
     */
    static _hasJavaNature(projectPath, src) {
        return hasJavaNature([path.join(projectPath, src), projectPath])
    }

    static _getForValueFromTask(task) {
        return task.for ? task.for : task.use && task.use.substring(BUILD_TASK_PREFIX.length + 1)
    }

    /**
     * For valid paths remove trailing '/'. Otherwise return as is - important!!
     * @param {*} dir
     */
    static _normalizePath(dir) {
        return typeof dir === "string" ? dir.replace(/\/$/, '') : dir
    }

    /**
     * Determines the module folder from the past list that may represent files or folders w or w/o .cds file extension.
     * @param {string} projectPath
     * @param {Array} filesOrFolders
     */
    static _getModuleFolder(projectPath, filesOrFolders) {
        const resources = [...filesOrFolders]
        filesOrFolders.forEach(fileOrFolder => {
            if (path.extname(fileOrFolder) !== FILE_EXT_CDS) {
                resources.push(fileOrFolder + FILE_EXT_CDS)
            }
        })
        return resources.reduce((acc, resource) => {
            if (!acc) {
                let resourcePath = path.resolve(projectPath, resource)
                if (fs.existsSync(resourcePath)) {
                    if (fs.lstatSync(resourcePath).isDirectory()) {
                        acc = resource
                    } else {
                        // represents file
                        acc = path.dirname(resource)
                    }
                }
            }
            return acc
        }, null)
    }
}
module.exports = BuildTaskProviderInternal
